### PR TITLE
Add new "dialogs" feature

### PR DESCRIPTION
**Pull request by reinink**
This PR adds first class support for dialogs (modals) to Inertia.js. It works by adding a dialog component to the DOM as a sibling to the page component. Dialogs have their own props and URL. To open a dialog, you simply navigate to that endpoint.

- [x] #928
- [x] #929
- [x] #930
- [x] #931
- [x] #932

### CODE REVIEW

File: packages/inertia/src/router.ts [hunk 1]
```diff
@@ -90,9 +90,14 @@ export class Router {
...
 
-  protected restoreScrollPositions(): void {
+  protected restoreScrollPositions(attempt = 1): void {
     if (this.page.scrollRegions) {
-      this.scrollRegions().forEach((region: Element, index: number) => {
+      const availableRegions = this.scrollRegions()
+      if (attempt <= 20 && (this.page.scrollRegions.length !== availableRegions.length)) {
+        setTimeout(() => this.restoreScrollPositions(attempt + 1), 2)
+        return
+      }
```
Comment (reinink) [inline packages/inertia/src/router.ts:99]: @claudiodekker Can we use `queueMicrotask` instead?

Or even `Promise.resolve().then(() => {})`

**Comment by fourstacks** [comment]
Hey @reinink, quick question about how best to build towards this dialogs feature for when it lands if you can spare the time?

Would I be correct in saying that dialogs will essentially take the content of the 'destination' page and display it within the slot of whatever dialogue component is specified?

So if I ultimately want to have post creation form appear in a modal but I want that post creation view to have it's own route/controller etc should I (for now) just build that form as a standalone page and when the dialoags feature ships it should mostly just work once I hook up the new API?

Thanks!

**Comment by reinink** [comment]
> So if I ultimately want to have post creation form appear in a modal but I want that post creation view to have it's own route/controller etc should I (for now) just build that form as a standalone page and when the dialoags feature ships it should mostly just work once I hook up the new API?

@fourstacks Yup, that's pretty much it!

**Comment by edwingromero** [comment]
Hello I am about to start a new project where many modal/dialog ways. And I need to click on a route to open a page and modal.

How long will this feature be available?

**Comment by maxeckel** [comment]
@edwingromero I've read somewhere, that Jonathan hopes or wants to showcase the feature on Laracon, but that won't mark a release or something.

Until then, there is a great write up by Lars Klopstra about how you can achive this in your app (vue-based).

https://larsklopstra.nl/posts/route-based-modals-with-inertiajs-and-vuejs

**Comment by edwingromero** [comment]
> @edwingromero I've read somewhere, that Jonathan hopes or wants to showcase the feature on Laracon, but that won't mark a release or something.
> 
> Until then, there is a great write up by Lars Klopstra about how you can achive this in your app (vue-based).
> 
> https://larsklopstra.nl/posts/route-based-modals-with-inertiajs-and-vuejs

Thanks for the answer, I will review the article

**Comment by dangerdan** [comment]
Is there a status for this? It's such an exciting feature, and will solve a lot of issues I'm trying to work around.

**Comment by claudiodekker** [comment]
@dangerdan we're actively working on it :)

**Comment by geidelguerra** [comment]
@claudiodekker @reinink I there a way I can help you with this feature guys?

**Comment by oliverbj** [comment]
Would also like to offer my help - really excited for this. It's like the last mizzing piece of the puzzle 🥇

**Comment by ctessier** [comment]
Hi @reinink @claudiodekker 

I am also very interested in this feature. I am a React developer. How can we help to ship this feature asap ? 
I just watched the demo at last month's Laracon. It looks AMAZING! Awesome job 👏👏

Let us know 😉
Thank you

**Comment by tpetry** [comment]
Watched the Laracon presentation again after being mentioned by @ctessier. Having the ability to define a page the dialog is being used on is very nice! But will there be a solution for the common problem of having stacked modals?

Like for the ping crm demo application you are on the users page. You hit "Create New User" and don't find the organization you want to use in the dropdown. So instead of closing the dialog, going to the organizations page add it and go back to the users page to open the modal again you are just click on a button within the dialog to create a new organization. The modal is stacked on-top of the other one (or hides the other one) and you create it the organization then you have the new organization in the dropdown list. The standard problem of forms needing linked relationships. In [wire-elements/modal](https://github.com/wire-elements/modal) this is named child modals.

**Comment by cwbmuller** [comment]
@claudiodekker @reinink Just checking if there is any update regarding the dialogs feature and its expected timelines 🙏

**Comment by cnb-philip** [comment]
Currently rolling my own hacky workaround version of this feature, as our new design makes use of tons of modals/drawers.  I don't want my team to have to sometimes use inertia forms, and then sometimes standard axios + non-inertia forms.  Would love to see this feature sooner than later.

**Comment by oliverbj** [comment]
> Currently rolling my own hacky workaround version of this feature, as our new design makes use of tons of modals/drawers. I don't want my team to have to sometimes use inertia forms, and then sometimes standard axios + non-inertia forms. Would love to see this feature sooner than later.

Exactly same situation. I'm actually holding of starting a new internal project in my organization, because it will also rely on the dialogs, and I don't want to create my own version if this feature is coming any time soon. 😅

**Comment by matthewknill** [comment]
> > Currently rolling my own hacky workaround version of this feature, as our new design makes use of tons of modals/drawers. I don't want my team to have to sometimes use inertia forms, and then sometimes standard axios + non-inertia forms. Would love to see this feature sooner than later.
> 
> Exactly same situation. I'm actually holding of starting a new internal project in my organization, because it will also rely on the dialogs, and I don't want to create my own version if this feature is coming any time soon. 😅

We are also holding off starting a new project in anticipation of this feature. Super keen for a release that has dialogs!

**Comment by ctessier** [comment]
> Is this going to be replaced by the stacking feature as discussed by @reinink here https://twitter.com/reinink/status/1376622208903225349 ?
> 
> Any outlook on a timetable?

Man, you're confusing me 😄 This tweet is more than a year old (march 2021), and Jonathan did a demo of the dialog feature, back at Laracon Online in summer 2021, and there was nothing about "stacking" or whatsoever. 
Although, in the tweet, he was still talking about "modals" and not "dialog", which he said in the demo that the latter is a more suitable term for the feature.

So, I am actually wondering if the "stacking" is still in the picture... The demo with the dialog seemed pretty neat and stable to me.

**Comment by iraklisg** [comment]
Hello. Is there an estimate on whether this feature is going to be released? Thank you

**Comment by aviemet** [comment]
The implementation looks very elegant, looking forward to its release. I have a question about the `baseRoute` option.

In my application I have standard pages for "create" forms. For related models on these forms, I'll have a "new" button to display a  form for the related record in a modal dialog. This basically means the dialog route wouldn't have a useful base route since I would only ever display this as a modal if it was visited from within a form for another record. 

So, for instance, `/items/new` would have a link to `/vendors/create` to show a modal, but so would `/accessories/new`, etc. This means `/vendors/create` doesn't have one single route or path which makes sense to set as a base route. There is however, a `/vendors/new` route which renders a standard inertia page with the form on it, no modal.

Will it be possible to conditionally redirect to another page if a dialog is visited from outside the context of a modal? So in the example above, visiting the dialog route `/vendors/create` directly would instead redirect to `/vendors/new`. 

Or perhaps somehow include the base page in the url so the base route is fetched dynamically from request params? I'm thinking something like `/items/new/vendors/create` which would render a base page of `/items/new` and also call the dialog route for `/vendors/create`.

Thank you

**Comment by Frozire** [comment]
Since no one has linked this here yet, https://github.com/lepikhinb/momentum-modal provides an excellent alternative for this until further progress on the feature here. The author has even tried to stay near this implementation to have an easy migration in the future.

**Comment by Tofandel** [comment]
Looks awesome and almost what I had in mind, does it support nested modals already?

**Comment by sbc640964** [comment]
Can I take a loan on this feature?
Or maybe build some kind of building on it?
Or not to bet for now?
Is there a plan? Is there a prospect? Make balloons and candles for a party?

**Comment by aviemet** [comment]
> Since no one has linked this here yet, https://github.com/lepikhinb/momentum-modal provides an excellent alternative for this until further progress on the feature here. The author has even tried to stay near this implementation to have an easy migration in the future.

Do you happen to know if there is something like momentum for Rails inertia apps?

**Comment by marktan93** [comment]
> > Since no one has linked this here yet, https://github.com/lepikhinb/momentum-modal provides an excellent alternative for this until further progress on the feature here. The author has even tried to stay near this implementation to have an easy migration in the future.
> 
> Do you happen to know if there is something like momentum for Rails inertia apps?

I don't think there is any package ready for rails.
I used reactjs and I took the momentum PHP code and self-wrote the frontend side.

**Comment by sbc640964** [comment]
> I used reactjs and I took the momentum PHP code and self-wrote the frontend side.

@marktan93 Can you share your React code?

**Comment by ThaDaVos** [comment]
Any progress on this? Just ran into a scenario where I need this, after digging around found this issue and saw the last commit is from April 2021

**Comment by marktan93** [comment]
> > I used reactjs and I took the momentum PHP code and self-wrote the frontend side.
> 
> @marktan93 Can you share your React code?

i just used the momentum-modal package and then self create a custom modal provider at react js 

`import { Inertia } from "@inertiajs/inertia";
import React, { useState, createContext, useContext } from "react";

const ModalContext = createContext<any>("");

type Modal = {
    component: string;
    baseURL: string;
    redirectURL: string | null | undefined;
    props: Record<string, any>;
    key: string;
    nonce: string;
} | null;

export const ModalProvider = ({ children }: any) => {
    const [show, setShow] = useState(false);
    const [redirect, setRedirect] = useState("");
    const [title, setTitle] = useState("");
    const [modal, setModal] = useState<Modal>(null);

    return (
        <ModalContext.Provider
            value={{
                show,
                setShow,
                redirect,
                setRedirect,
                title,
                setTitle,
                modal,
                setModal,
            }}
        >
            {children}
        </ModalContext.Provider>
    );
};

export const useModal = () => {
    const {
        show,
        setShow,
        redirect,
        setRedirect,
        title,
        setTitle,
        modal,
        setModal,
    } = useContext(ModalContext);

    const ModalComponent = React.lazy(
        () => import("./../Pages/" + modal?.component)
    );

    const close = () => {
        setShow(false);
        setModal(null);

        return Inertia.visit(modal?.redirectURL, {
            preserveScroll: true,
            preserveState: true,
        });
    };

    return {
        ModalComponent,
        show,
        setShow,
        redirect,
        setRedirect,
        title,
        setTitle,
        modal,
        setModal,
        close,
    };
};
`

**Comment by dwainscheeren** [comment]
I just saw the video about the new dialog thing from last year and I wanted to switch from livewire to inertiajs, but now I see it is not yet released. For my projects I really need dialog/modals to create and edit rows. Any idea when this will be released?

**Comment by kenaths** [comment]
> > > I used reactjs and I took the momentum PHP code and self-wrote the frontend side.
> > 
> > 
> > @marktan93 Can you share your React code?
> 
> i just used the momentum-modal package and then self create a custom modal provider at react js
> 
> ```
> import { Inertia } from "@inertiajs/inertia";
> import React, { useState, createContext, useContext } from "react";
> 
> const ModalContext = createContext<any>("");
> 
> type Modal = {
>     component: string;
>     baseURL: string;
>     redirectURL: string | null | undefined;
>     props: Record<string, any>;
>     key: string;
>     nonce: string;
> } | null;
> 
> export const ModalProvider = ({ children }: any) => {
>     const [show, setShow] = useState(false);
>     const [redirect, setRedirect] = useState("");
>     const [title, setTitle] = useState("");
>     const [modal, setModal] = useState<Modal>(null);
> 
>     return (
>         <ModalContext.Provider
>             value={{
>                 show,
>                 setShow,
>                 redirect,
>                 setRedirect,
>                 title,
>                 setTitle,
>                 modal,
>                 setModal,
>             }}
>         >
>             {children}
>         </ModalContext.Provider>
>     );
> };
> 
> export const useModal = () => {
>     const {
>         show,
>         setShow,
>         redirect,
>         setRedirect,
>         title,
>         setTitle,
>         modal,
>         setModal,
>     } = useContext(ModalContext);
> 
>     const ModalComponent = React.lazy(
>         () => import("./../Pages/" + modal?.component)
>     );
> 
>     const close = () => {
>         setShow(false);
>         setModal(null);
> 
>         return Inertia.visit(modal?.redirectURL, {
>             preserveScroll: true,
>             preserveState: true,
>         });
>     };
> 
>     return {
>         ModalComponent,
>         show,
>         setShow,
>         redirect,
>         setRedirect,
>         title,
>         setTitle,
>         modal,
>         setModal,
>         close,
>     };
> };
> ```



> marktan93

@marktan93 Have you made any changes to app.ts?

**Comment by jaulz** [comment]
(shameless plug) If anyone is still looking for a solution for Vue3 that supports modals, you can find something similar to Inertia here:
https://github.com/navigarejs/framework 

It grew on top of Inertia's code base and documentation but unfortunately it is not 100% compatible. However, it comes with a lot of features:
- Complex layouts with persistent parts (e.g. headers) or stacked fragments (e.g. modals)
- Preloading of links
- Route injection without any client overhead
- (Optional) real-time form validation via Precognition
- Built-in (optional) SSR support from the beginning

**Comment by Snicser** [comment]
Is there already a release date for this feature?

**Comment by dwainscheeren** [comment]
Is there already a release date for this feature?

**Comment by ZayRTun** [comment]
Is this releasing any time soon? I really need this feature...

**Comment by taozuhong** [comment]
Maybe expose the axios to communicate with server with shared session, then more solutions will be made.

**Comment by nagi1** [comment]
What are features that need to be completed in order to merge this PR?

**Comment by Dibbyo456** [comment]
Hi, do anyone have slightest idea why this feat was abandoned?

**Comment by ctessier** [comment]
Hi, @reinink 👋 Can we have just one reply please? 🥺 So we can at least get an idea whether this feature will be published or not ever? Thank you so much!

**Comment by Dibbyo456** [comment]
There is a third-party package [lepikhinb/momentum-modal](https://github.com/lepikhinb/momentum-modal), add the modal feature. But the frontend plugin [lepikhinb/momentum-modal-plugin](https://github.com/lepikhinb/momentum-modal-plugin) available for Vue3 only.

**Comment by baoanhng** [comment]
Very promising but sadly. Delay is expected for every web app out there.

**Comment by rvetrsek** [comment]
Why was this feature announced and even presented as part of a YouTube video 2 years ago and was never released? This is the missing piece that is keeping me from using Inertia entirely in all projects.

Am I missing something, or is there no way to manage modals with history and direct linking without this?

**Comment by rvetrsek** [comment]
Well, @Dibbyo456, thank you, but I'm using React, unfortunately.

**Comment by willianantunes** [comment]
Have you seen the following project? It supports React. 😁 

https://github.com/inertiaui/modal

**Pull request closed without merging**